[SR-10374] URLCache: init method and first time sqlite database setup implemented

karthikkeyan opened the pull request:
* `init` method implemented
* `URLCache.shared` singleton object created with 4MB of memory space and 20MB of disk space
* Directory and database file Cache.db file created under the local directory
* Sqlite Tables and Indices created in the database
* Unit tests added for `URLCache` to verify directory, file, tables, and indices

Bug Link: https://bugs.swift.org/browse/SR-10374

pushkarnk commented:
@karthikkeyan Apologies for the very late response on this PR. I think implementations of URLCache and URLCredentialStorage have been long pending and this looks like a great start! However, I've a few questions about this pull request:
1. Is SQLite available on Linux and Windows? I guess not.
2. Have we considered other storage alternatives? 
I came across this [old discussion on URLCache ](https://forums.swift.org/t/implementing-urlcache/13398). You might find it relevant.

karthikkeyan commented:
> @karthikkeyan Apologies for the very late response on this PR. I think implementations of URLCache and URLCredentialStorage have been long pending and this looks like a great start! However, I've a few questions about this pull request:
> 
> 1. Is SQLite available on Linux and Windows? I guess not.
> 2. Have we considered other storage alternatives?
>    I came across this [old discussion on URLCache ](https://forums.swift.org/t/implementing-urlcache/13398). You might find it relevant.

I have failed to give a fair bit of a thought process about the Linux platform. Thanks for reminding of it, I will follow the discussion and see how best we can move this PR forward.

karthikkeyan commented:
@pushkarnk I have reimplemented store and retrieve logic using `NSKeyedArchiver` and `NSKeyedUnarchiver`. This implementation is not perfect but I think it would be a good start.

spevans commented:
@swift-ci test

drodriguez commented on Foundation/URLCache.swift:
```diff
@@ -142,13 +147,37 @@ open class URLCache : NSObject {
+            
+            sharedSyncQ.sync {
+                // Remove previous data before resetting new URLCache instance
+                URLCache.sharedCache?.persistence?.deleteAllCaches()
```
Does the original implementation do this? Feels weird removing “user” data (even if it is a cache), just by assigning another value.

millenomi replied:
It does not.

drodriguez commented on Foundation/URLCache.swift:
```diff
@@ -243,11 +283,91 @@ open class URLCache : NSObject {
+
+    let path: String
+
+    var cacheSizeInDesk: Int {
```
Typo: `cacheSizeInDisk`

drodriguez commented on Foundation/URLCache.swift:
```diff
@@ -243,11 +283,91 @@ open class URLCache : NSObject {
+                let attributes = try FileManager.default.attributesOfItem(atPath: path.appending(fileName))
+                total += (attributes[.size] as? Int) ?? 0
+            }
+            return total
```
Is there plan to keep this value updated as cached responses are added and evicted? Feels like with a big cache and a slow disk, each of these calls might take a while.

karthikkeyan replied:
Let me add a `FIXME` comment, I will create a separate PR for optimizing this. My main focus in this PR is to create an initial setup and kick start the implementation. Hope that's ok.

drodriguez commented on Foundation/URLCache.swift:
```diff
@@ -243,11 +283,91 @@ open class URLCache : NSObject {
+    func setupCacheDirectory() {
+        do {
+            if FileManager.default.fileExists(atPath: path) {
+                try FileManager.default.removeItem(atPath: path)
```
Wouldn’t this delete the previous cache contents?

millenomi replied:
Indeed.

drodriguez commented on Foundation/URLCache.swift:
```diff
@@ -243,11 +283,91 @@ open class URLCache : NSObject {
+    }
+
+    private func urlForFileIdentifier(_ identifier: String) -> URL {
+        return URL(fileURLWithPath: path + "/" + identifier)
```
It would be good to use `URL(fileURLWithPath: path, isDirectory: true).appendingPathComponent(identifier)` to handle platforms with other directory separators.

drodriguez commented on Foundation/URLCache.swift:
```diff
@@ -243,11 +283,91 @@ open class URLCache : NSObject {
+        }
+
+        let method = httpMethod ?? "GET"
+        return urlString + method
```
Since this is used as part of the path, shoulnd’t the output be escaped or hashed somehow?

millenomi replied:
Indeed.

karthikkeyan commented:
> Full disclosure: I didn’t know about this PR, and I started working on my own `URLCache` (that's the reason for #2245). I have some parts that are present in this version (with some differences), but I focused in the memory cache first. I have ideas for the disk cache, but that's more complicated. I can also have a lot more feedback about things I know that will need to be handle (for example skipping some protocols from being cached). Are you still planning in working on this?

@drodriguez  Sorry for the delay response, yes I will be working on this PR. I was just waiting for someone to review this. Please give me your suggestions.

drodriguez commented on Foundation/URLCache.swift:
```diff
@@ -142,13 +147,37 @@ open class URLCache : NSObject {
+                } else {
+                    let fourMegaByte = 4 * 1024 * 1024
+                    let twentyMegaByte = 20 * 1024 * 1024
+                    let bundleIdentifier = Bundle.main.bundleIdentifier ?? UUID().uuidString
```
That `UUID` at the end is a little bit worrying. The app will fill caches, and will never reuse the cached results, filling the disk of garbage. I can think at least of `ProcessInfo.processInfo.processName` as a good fallback, but it is not unique (but the bundle identifier doesn’t have to be either). Maybe if no bundle identifier is found the best thing is not to have a disk cache at all.

millenomi replied:
Can this point to the temporary directory with a fixed name? (A sanitized process name does sound good to me!)

drodriguez commented on Foundation/URLCache.swift:
```diff
@@ -142,13 +147,37 @@ open class URLCache : NSObject {
+                    let fourMegaByte = 4 * 1024 * 1024
+                    let twentyMegaByte = 20 * 1024 * 1024
+                    let bundleIdentifier = Bundle.main.bundleIdentifier ?? UUID().uuidString
+                    let cacheDirectoryPath = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first?.path ?? "\(NSHomeDirectory())/Library/Caches"
```
Same idea here. If we don’t find a good URL for `.cachesDirectory` maybe it is better to not have disk cache.

The fallback has the problem of being very Darwin specific. For example `NSHomeDirectory()` in Android is the root directory, where one cannot write at all. Also, it might want to use `appendingPathComponent(…).appendingPathComponent(…)` to be nicer to Windows.

(There's another case below where the POSIX path separator is used instead of a higher level function).

millenomi replied:
The home directory should be in a persistent writable path everywhere, much as possible. I know that Android is _really_ weird about this, but that looks like incorrect mapping to me.

millenomi replied:
Yeah, please do not fall back like this. If `FileManager` does not return a caches URL, consider not caching at all.

karthikkeyan replied:
Sure, I will remove the fallback if `FileManager` for `cachesDirectory` returned nill.

drodriguez commented on Foundation/URLCache.swift:
```diff
@@ -162,7 +191,16 @@ open class URLCache : NSObject {
+        self.memoryCapacity = memoryCapacity
+        self.diskCapacity = diskCapacity
+
+        if let _path = path {
```
You might also want to check that `diskCapacity` is more than zero.

millenomi replied:
Darwin indeed does accept 0 arguments. My assumption is that memory capacity 0 means that as much as possible no data is kept in memory, and disk capacity 0 that no data is written to disk.

drodriguez commented on Foundation/URLCache.swift:
```diff
@@ -243,11 +283,91 @@ open class URLCache : NSObject {
+
+fileprivate extension URLRequest {
+
+    var cacheFileIdentifier: String {
```
I have a slightly different design here (I use a `URLCacheKey` type, instead of `String`), but one important thing is that generating a key from the `URLRequest` depends on some details from the request beside the URL itself. So my equivalent method returns an optional.

The first thing is the protocol. Protocols like `data:` or `file:` should not be cached for obvious reasons. I also think unknown protocols should not be cached either.

Also, for known protocols like `http:` or `https:` (and maybe for `ftp:`, but I don’t know that protocol that well), we should do a couple more things. For example:
- I don’t think you can do a HTTP request without a method, so that's a `return nil` in the case the protocol is HTTP.
- I think some headers should be taken into account. The spec refers to `User-Agent`, but others like `Accept-Language` and `Range` might be interesting.
- I think the hash of the body contents should also be used.
- I also think one should not try to cache request with body streams, because that will consume it, which is not good.
- Finally I would hash everything and return an hexadecimal string, to have a valid file name.

millenomi commented:
I want to shepherd this patch into the repository. Do you need help with changes here?

karthikkeyan commented:
> I want to shepherd this patch into the repository. Do you need help with changes here?

@millenomi I am addressing all the comments added by you and Danial. I will let you know if I need any help. Thanks Lily.

karthikkeyan commented:
@millenomi @drodriguez I have addressed all the comments. I still have some doubts about the implementation of the property `cacheFileIdentifier`, please let me know if there is any changes needed.

millenomi commented:
@swift-ci please test

millenomi commented on Foundation/URLCache.swift:
```diff
@@ -248,11 +289,101 @@ open class URLCache : NSObject {
+            return nil
+        }
+
+        var hashString = "\(scheme)_\(method)_\(urlString)"
```
The collision space of the default hash is smaller than you probably think. Can we at least check when we load a request that this is actually the request we were looking for?

millenomi commented on Foundation/URLCache.swift:
```diff
@@ -248,11 +289,101 @@ open class URLCache : NSObject {
+
+        let url = urlForFileIdentifier(fileIdentifier)
+        guard let data = try? Data(contentsOf: url),
+            let response = NSKeyedUnarchiver.unarchiveObject(with: data) as? CachedURLResponse else {
```
Please use secure coding here.

millenomi replied:
To use secure coding, create a `NSKeyedUnarchiver`, set its `requiresSecureCoding` to `true`, and use its method to encode an object for key `NSKeyedArchiveRootObjectKey`.

millenomi commented on Foundation/URLCache.swift:
```diff
@@ -248,11 +289,101 @@ open class URLCache : NSObject {
+    }
+
+    func saveCachedResponse(_ response: CachedURLResponse, for request: URLRequest) {
+        let archivedData = NSKeyedArchiver.archivedData(withRootObject: response)
```
This is slightly less urgent, but I'd like you to use secure coding wherever you use coding.

millenomi commented on Foundation/URLCache.swift:
```diff
@@ -248,11 +289,101 @@ open class URLCache : NSObject {
+        do {
+            try FileManager.default.createDirectory(atPath: path, withIntermediateDirectories: true)
+        } catch {
+            fatalError("Unable to create directories for URLCache: \(error.localizedDescription)")
```
Do not introduce fatal errors for things the user may not be able to control. As I asked elsewhere: consider not loading if we do not have a suitable directory.

millenomi commented:
I'm going to commandeer this patch; I may drop the SQLite dependency, though.

karthikkeyan commented:
> I'm going to commandeer this patch; I may drop the SQLite dependency, though.

@millenomi  I have pushed one more commit addressing the comments related to secure archiving. Also, I have updated the hashing mechanism for generating `cacheFileIdentifier`. 

The one thing I am struggling though is to write a unit test for the internal property `cacheFileIdentifier`

karthikkeyan commented:
Closing this PR as `URLCache` is implemented in this PR https://github.com/apple/swift-corelibs-foundation/pull/2401
